File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    /*
        bfs로 들어가서 깊이를 구하는게 좋을 듯 하다.
    */
    
    static class Node {
        int node;
        int depth;
        
        Node(int node, int depth) {
            this.node = node;
            this.depth = depth;
        }
    }
    
    static Map<Integer, Set<Integer>> graph;
    
    static int maxLen = 0;
    
    static int cnt = 0;
    
    public int solution(int n, int[][] edge) {
        int answer = 0;
        
        graph = new HashMap<>();
        
        for(int i = 0; i <= n; i++) {
            graph.put(i, new HashSet<>());
        }
        for(int i = 0; i < edge.length; i++) {
            graph.get(edge[i][0]).add(edge[i][1]);
            graph.get(edge[i][1]).add(edge[i][0]);
        }
        
        bfs();        
        
        answer = cnt;
        
        return answer;
    }
    
    public static void bfs() {
        int[] visited = new int[graph.size()];
        Queue<Node> q = new LinkedList<>();
        
        visited[1] = 1;
        q.add(new Node(1, 0));
        
        while(!q.isEmpty()) {
            Node nn = q.poll();
            
            if(maxLen < nn.depth) {
                maxLen = nn.depth;
                cnt = 1;
            } else if (maxLen == nn.depth) {
                cnt++;
            }
            
            for(Integer num: graph.get(nn.node)) {
                if(visited[num] == 1) continue;
                
                visited[num] = 1;
                q.add(new Node(num, nn.depth + 1));
            }
        }
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    var answer = 0;
    
    // Init variable
    // 연결 여부를 2차원 배열로 저장하면 메모리 초과
    let q = []
    let vis = Array(n+1).fill(0)
    let maxCnt = 0
    
    // let adjList = Array(n+1).fill().map(() => new Set())
    let adjList = Array(n+1).fill().map(() => [])
    // let adjList = Array.from(Array(n+1), () => [])
    edge.forEach(([u, v]) => {
        adjList[u].push(v)
        adjList[v].push(u)
    })
    console.log(adjList)
    
    // bfs
    vis[1] = 1
    q.push([1, 0])
    
    while(q.length !== 0) {
        const [node, cur_count] = q.shift()
        
        if(maxCnt < cur_count) {
            maxCnt = cur_count
            answer = 1
        } else if(maxCnt === cur_count) {
            answer++
        }
        
        adjList[node].forEach((nextNode) => {
            if(vis[nextNode] === 1) return

            // console.log("node=", nextNode)
            // console.log("vis=", vis)
            
            vis[nextNode] = 1
            q.push([nextNode, cur_count+1])  
        })
    }
    
    return answer;
}